actions/mouths-of-watersheds/mouths-of-watersheds.py

```diff
@@ -0,0 +1,52 @@
+import argparse
+import csv
+from netCDF4 import Dataset
+from shapely.geometry import shape
+import geojson
+from mow.helpers import *
+from mow.geo_data_grid_2d.vic import VicDataGrid
+
+parser = argparse.ArgumentParser('Output a CSV file with the lon lat of the mouth of each watershed')
+parser.add_argument('nc', help='an RVIC-netCDF file')
+parser.add_argument('geojson', help='a geoserver output file describing the watersheds')
+parser.add_argument('csv', help='the csv file to output the results to')
+args = parser.parse_args()
+nc = Dataset(args.nc, "r")
+flow_direction = VicDataGrid.from_nc_dataset(nc, "flow_direction")
+grid = VIC_direction_matrix(flow_direction.lat_step, flow_direction.lon_step) 
+
+with open(args.geojson) as f:
+    with open(args.csv, 'w', encoding='UTF8') as x:
+        writer = csv.writer(x)
+        header = ['lon', 'lat', 'properties']
+        writer.writerow(header)
+        gj = geojson.load(f)
+        for i in gj['features']:
+            visited = set()
+            center = shape(i.geometry).centroid
```

Comment:
Th centroid is not guaranteed to be inside the polygon unless it is convex (and watersheds often are not). James suggests using [`representative_point()`](https://shapely.readthedocs.io/en/stable/manual.html#object.representative_point), which is guaranteed to be so.

Reply:
I have reservations about using `representative_point()`.  I originally wanted to use `centroid()` because we're building a correspondence between two systems (flow netCDF and the polygone geoJSON) that have different resolutions - the polygons are higher resolution than the flow map. I wanted a point near the center because I think the two systems may disagree about whether a particular points near the boundaries are in the watershed, but won't disagree about a point near the center.

Reply:
I guess one option would be to try `centroid()`, and if that fails, try `representative_point()`, but raise a warning so the user knows to double-check that one.

Reply:
There is, unfortunately, at least one watershed whose centroid is not in the watershed. Behold, the Lower Halfway River watershed. 
![lower_halfway](https://user-images.githubusercontent.com/4512605/164507852-3a274646-e6c4-4631-a88d-abb01b52e02b.png)

Reply:
Looks like there are 5-6 such watersheds in BC.

Reply:
When I tried to use ```representative_point```, there were actually fewer watersheds with points inside of the netCDF file (19 vs 20) and one of the watersheds returned a different mouth than before. And when using ```representative_point``` and removing my check for a valid point, I get the same error, as some of the points are still on the masked values.

Reply:
So the have the following problems here.
1. Centroid fails for some watersheds because concavity.
2. `representative_point` fails on some watersheds because of boundary mismatches.

There may be a third problem to do with masked values, but I think hitting a masked value, however you obtained the point, means that your point is outside the flow direction grid.

If that's true, a possible solution is:
- use `representative_point`
- if a masked value (outside of flow grid), then search within a small distance (in cell indexes) of the masked point for a location that is (a) not masked (inside flow grid) and (b) inside the polygon of interest.

Reply:
As a final sanity check, it would be worth verifying that each mouth cell computed overlaps its watershed polygon (or is very very very near it).

Reply:
>  and one of the watersheds returned a different mouth than before.

If we're not sure that our original answers were strictly correct, then I wouldn't necessarily see that as a problem.

Can you map the watersheds + the result of `representative_point()` for watersheds where the point is outside of the flow direction grid? That will likely inform our solution.

Reply:
After figuring out an error in my code, I now am getting unique mouths for each watershed. The ```representative_point``` 19 values are identical to their corresponding ```centroid``` values. (of which there are 20). The only difference between the two outputs is that code ```WTRSHDGRPD 25``` is not in the netCDF when using ```representative_point```

Reply:
> @corviday , formally, they're all concave, but, by eye, centroid ought to produce a point inside the watershed, except maybe the 4th one. Puzzling.

For all except maybe the fourth, and possibly also the fourth, they're probably just in a different file. Each file is one river, with watersheds that drain into different rivers masked off. But since files are square and rivers are squiggly, the spatial extent of each river-file will encompass a few watersheds that drain into a different river.

> What we do from here depends on what the long-term goal for this script is. If this is a one-time shot (and in my experience, there very very few one-time shots, even when you think it is), then perhaps just leaving it where it lies is okay.

This script will be used to make canned data that allows us to translate between the province's official watershed definitions (polygons) and the way PCIC models watersheds for calculations (grid-based, each watershed is defined as all grid squares upstream from a specific point, the "outlet" or "mouth"). 

One we have a calculated outlet for each of BC's official watersheds, we'll be able to use the PCIC data files to calculate data about the official watersheds. It's approximately one-time, at least, though I imagine it will be reused as Hydrology models or updates the gridded flow files - we don't have all BC  modeled yet.